File: 2015-01-LuhnTest/luhn_sigmundv.hs
-- Run as: runhaskell luhn.hs

import Data.Char (digitToInt)

-- First we convert the given integer to a list of integers

toDigits :: Int -> [Int]
toDigits n
	| length (show n) > 0 = digitToInt(head (show n)) : [digitToInt(i) | i <- tail (show n)]
	| otherwise = []

-- Then we define a helper function for reversing the given integer
	
toDigitsRev :: Int -> [Int]
toDigitsRev n = reverse (toDigits n)

-- Now it's time to define the function that doubles every second digit of the integer

doubleSecond :: [Int] -> [Int]
doubleSecond [] = []
doubleSecond [x] = [x]
doubleSecond (x0:x1:xs) = x0 : (2*x1) : doubleSecond xs

-- Before we combine everything into one function that checks if the credit card number is valid
-- we define a function for adding up all the digits of the integer

sumDigits :: [Int] -> Int
sumDigits xs = sum (concat [toDigits n | n <- xs])

-- Finally we combine all the functions above into one function
-- that checks the validity of a given credit card number

isValid :: Int -> Bool
isValid n = mod (sumDigits (doubleSecond (toDigitsRev n) ) ) 10 == 0

main :: IO()
main = do
	print (map isValid [49927398716,49927398717,1234567812345678,1234567812345670])

File: 2015-01-LuhnTest/luhn_sigmundv.py
# Run as: python luhn.py

# First we define a function that converts an integer to a list of its digits

def toDigits(n):
	string = str(n)
	digits = map(int,string)
	return list(digits)
	
# Then we define a helper function that reverses said digits
	
def toDigitsRev(n):
	return toDigits(n)[::-1]
	
# Now we double every even numbered digit
# (since Python lists are 0-indexed it translates to every odd element here)
	
def doubleSecond(digits):
	return [2*digits[i] if i % 2 != 0 else digits[i]
							for i in range(len(digits))]

# This function ensures that any two-digit numbers left from the multiplications above
# are split into single digits.

def sumDigits(numbers):
	lists = list(map(toDigits, numbers) # convert each number to a list of digits
	joined = sum(lists,[]) # concatenate the lists
	return sum(concatenated)

# Finally the functions from above are combined to form a function
# that checks if a given integer is a valid credit card number
	
def isValid(n):
	return sumDigits( doubleSecond( toDigitsRev(n) ) ) % 10 == 0


if __name__ == '__main__':
	
	numbers = [49927398716,49927398717,1234567812345678,1234567812345670]
	
	print( list( map(isValid, numbers) ) )
